src/Data/Array.purs

```diff
@@ -1289,3 +1293,78 @@ unsafeIndex :: forall a. Partial => Array a -> Int -> a
 unsafeIndex = unsafeIndexImpl
 
 foreign import unsafeIndexImpl :: forall a. Array a -> Int -> a
+
+-- | Takes an arrays of length n and returns an array of Tuples with length n-1.
+-- | It's a "sliding window" view of this array with a window size of 2 and a step size of 1.
+-- |
+-- | ```purescript
+-- | sliding [1,2,3,4,5] = [(Tuple 1 2),(Tuple 2 3),(Tuple 3 4),(Tuple 4 5)]
+-- | ```
+-- |
+sliding :: forall a. Array a -> Array (Tuple a a)
+sliding l = zip l (drop 1 l)
+
+-- | Takes an arrays and returns an array of arrays as a "sliding window" view of this array.
+-- | Illegal arguments result in an empty Array.
+-- |
+-- | ```purescript
+-- | > import Data.Array (range)
+-- | > slidingSizeStep 3 2 (range 0 10) = [[0,1,2],[2,3,4],[4,5,6],[6,7,8],[8,9,10],[10]]
+-- | > slidingSizeStep 3 3 (range 0 10) = [[0,1,2],[3,4,5],[6,7,8],[9,10]]
+-- | ```
+-- |
+slidingSizeStep :: forall a. Int -> Int -> Array a -> Array (NonEmptyArray a)
+slidingSizeStep size step array 
+  | size <= 0 || step <= 0 = [] {- args not valid -}
+  | otherwise = 
+    let
+      maxIndex = (length array) - 1
+    in
+      rangeWithStep' 0 maxIndex step (\i -> NonEmptyArray (slice i (i + size) array))
+
+
+-- | Create an array containing a range of integers with a given step size, including both endpoints.
+-- | Illegal arguments result in an empty Array.
+-- |
+-- | ```purescript
+-- | > rangeWithStep 0 6 2 = [0,2,4,6]
+-- | > rangeWithStep 0 (-6) (-2) = [0,-2,-4,-6]
+-- | ```
+-- |
+rangeWithStep :: Int -> Int -> Int -> Array Int
+rangeWithStep start end step = rangeWithStep' start end step identity
+
+-- | Helper function to produce an array of elements like `rangeWithStep` with start, end and step, but immediatelly mapping over the result to work on one single mutable array.
+-- |
+-- | ```purescript
+-- | > rangeWithStep' 0 6 2 identity = [0,2,4,6]
+-- | > rangeWithStep' 0 6 2 (add 3) = [3,5,7,9]
+-- | ```
+rangeWithStep' :: forall t. Int -> Int -> Int -> (Int -> t) -> Array t
+rangeWithStep' start end step fn =
+  STA.run
+    ( do
+        let
+          isValid =
+            step /= 0
+              && if end >= start then
+                  step > 0
+                else
+                  step < 0
```

Comment:
Looks like `isValid` is no longer used in this function. Can this be removed?

Reply:
My bad - I forgot to use it again during my refactoring. It's still needed to validate the arguments if you want to have a range "in the negative direction".

Reply:
used it again and remembered, that purescript has the where block. Made it a bit more readable